crates/optimism/primitives/src/transaction/signed.rs

```diff
@@ -0,0 +1,440 @@
+        self.deref().ty()
+    }
+
+    fn access_list(&self) -> Option<&AccessList> {
+        self.deref().access_list()
+    }
+
+    fn blob_versioned_hashes(&self) -> Option<&[B256]> {
+        self.deref().blob_versioned_hashes()
+    }
+
+    fn authorization_list(&self) -> Option<&[SignedAuthorization]> {
+        self.deref().authorization_list()
+    }
+
+    fn is_dynamic_fee(&self) -> bool {
+        self.deref().is_dynamic_fee()
+    }
+
+    fn effective_gas_price(&self, base_fee: Option<u64>) -> u128 {
+        self.deref().effective_gas_price(base_fee)
+    }
+
+    fn effective_tip_per_gas(&self, base_fee: u64) -> Option<u128> {
+        self.deref().effective_tip_per_gas(base_fee)
+    }
+}
+
+impl Default for OpTransactionSigned {
+    fn default() -> Self {
+        Self {
+            hash: Default::default(),
+            signature: Signature::test_signature(),
+            transaction: OpTransaction::new(OpTypedTransaction::Legacy(Default::default())),
+        }
+    }
+}
+
+#[cfg(any(test, feature = "arbitrary"))]
+impl<'a> arbitrary::Arbitrary<'a> for OpTransactionSigned {
+    fn arbitrary(u: &mut arbitrary::Unstructured<'a>) -> arbitrary::Result<Self> {
+        #[allow(unused_mut)]
+        let mut transaction = OpTypedTransaction::arbitrary(u)?;
+
+        let secp = secp256k1::Secp256k1::new();
+        let key_pair = secp256k1::Keypair::new(&secp, &mut rand::thread_rng());
+        let signature = reth_primitives::transaction::util::secp256k1::sign_message(
+            B256::from_slice(&key_pair.secret_bytes()[..]),
+            signature_hash(&transaction),
+        )
+        .unwrap();
+
+        // Both `Some(0)` and `None` values are encoded as empty string byte. This introduces
+        // ambiguity in roundtrip tests. Patch the mint value of deposit transaction here, so that
+        // it's `None` if zero.
+        if let OpTypedTransaction::Deposit(ref mut tx_deposit) = transaction {
+            if tx_deposit.mint == Some(0) {
+                tx_deposit.mint = None;
+            }
+        }
+
+        let signature = if is_deposit(&transaction) { TxDeposit::signature() } else { signature };
+
+        let mut signed_tx = Self::new_unhashed(OpTransaction::new(transaction), signature);
+        if !signed_tx.tx_type().is_deposit() {
+            signed_tx.hash = signed_tx.recalculate_hash()
+        }
+
+        Ok(signed_tx)
+    }
+}
+
+/// Calculates the signing hash for the transaction.
+pub fn signature_hash(tx: &OpTypedTransaction) -> B256 {
+    match tx {
+        OpTypedTransaction::Legacy(tx) => tx.signature_hash(),
+        OpTypedTransaction::Eip2930(tx) => tx.signature_hash(),
+        OpTypedTransaction::Eip1559(tx) => tx.signature_hash(),
+        OpTypedTransaction::Eip7702(tx) => tx.signature_hash(),
+        OpTypedTransaction::Deposit(_) => B256::ZERO,
+    }
+}
+
+/// Returns `true` if transaction is deposit transaction.
+pub const fn is_deposit(tx: &OpTypedTransaction) -> bool {
```

Comment:
wh are those not functions of the type?

Reply:
you mean why that isn't in op-alloy? can open an issue for it but shouldn't block here

Reply:
https://github.com/alloy-rs/op-alloy/issues/315